File: leetcode/1501-1800/1651-1700/1678.Goal Parser Interpretation.cc
// https://leetcode.com/problems/goal-parser-interpretation/

class Solution {
public:
    string interpret(const string& command) const noexcept {
        string output;
        unsigned len = command.size();
        for (int i = 0; i < len; ++i) {
            if (command[i] == 'G') {
                output += 'G';
            } 
            else if (i - len > 1 && command[i] == '(') {
                if (command[i + 1] == ')') {
                    output += 'o';
                    i += 1;
                } 
                else if (command[i + 1] == 'a') {
                    output += "al";
                    i += 2;
                }
            }
        }
        
        return output;
    }
};

File: leetcode/1501-1800/1651-1700/1678.Goal Parser Interpretation.rs
// https://leetcode.com/problems/goal-parser-interpretation

impl Solution {
    pub fn interpret(command: String) -> String {
        command.replace("()","o").replace("(al)", "al")
    }
}

impl Solution {
    pub fn interpret(command: String) -> String {
        let mut result = String::new();
        let mut it = command.chars();
        let mut is_alpha = false;
        while let Some(c) = it.next() {
            let str = match c {
                'G' => "G",
                '(' => {
                    is_alpha = false;
                    continue                   
                },
                'a' => {
                    is_alpha = true;
                    "al"
                },
                ')' => if is_alpha { continue } else {is_alpha = false; "o"},
                _ => continue
            };

            result.push_str(&str.to_string());
        }

        result
    }
}